### PR TITLE
[bouncy] add package.xml patch for bouncy

### PR DESCRIPTION
**Pull request by mikaelarguedas**
this adds dependencies on all ROS2 interface packages we want to compile the bridge for:

 - actionlib_msgs
 - diagnostic_msgs
 - example_interfaces
 - geometry_msgs
 - nav_msgs
 - rosgraph_msgs
 - sensor_msgs
 - shape_msgs
 - std_srvs
 - stereo_msgs
 - tf2_msgs
 - trajectory_msgs
 - visualization_msgs

Dependencies already listed in source package.xml:
 - builtin_interfaces
 - std_msgs

### CODE REVIEW

**Comment by dirk-thomas** [comment]
Could these dependencies be put upstream? For from-source builds the package already depends on every message package using the group dependency. So I don't see how adding the additional packages (with a comment) would make any difference.

**Comment by mikaelarguedas** [comment]
Because users building from source should be able to build a dynamic bridge without having all these packages available.
At least that's been the hard restriction we've decided on in the past that's why we list only the packages that are *required* to build the bridge.

**Comment by dirk-thomas** [comment]
What do you mean with "without having all these packages available"?

If the packages are in the workspace they are being processed before because of the group dependency. If they are not in the workspace neither the group dependency nor the explicit dependency make any difference.

**Comment by nuclearsandwich** [comment]
> If the packages are in the workspace they are being processed before because of the group dependency. If they are not in the workspace neither the group dependency nor the explicit dependency make any difference.

Adding each message package to the source manifest individually when the group dependency already includes them seems redundant and requires extra work to maintain when message packages are added or removed. I'm not clear on what makes that extra work worthwhile.

The need to add each package manually during the release phase is required because rosdep doesn't have support for group dependencies. Is supporting the installation of all these packages from rosdep at source time a goal of the request to upstream the individual package dependencies?

**Comment by mikaelarguedas** [comment]
I mean that as a user I rely on the information in the package.xml to allow rosdep to install the dependencies *required* to build/use a package.
Adding these dependencies explicitly in the package.xml means that users will default to build a bigger bridge for messages that are not required and that they may not use.

As pointed out above now that we have the group dependencies, listing these dependencies in the source package.xml will duplicate information without changing behavior for the source case.

I don't feel strongly either way, though it seems that the better long term outcome (but not feasible for Bouncy) would be to have support for group dependencies in the non source case allowing us to keep the source package.xml as is and not have to patch it in the release repo.

We discussed where to list these dependencies the last few releases, for Ardent the conclusion was:

```
dthomas [8:24 PM]
the package.xml doesn't explicitly list them since it doesn't "require" those - the goal is to declare that using https://github.com/ros2/design/pull/138 in the future - therefore I am leaning to the patch in the release process
```

So I figured we'd keep going that route for Bouncy but I'm open to a different approach.

Could you elaborate on the motivation to have these in the source package.xml for Bouncy?
- Is it to avoid having any patch for this package in the debian case?
- Do we want users to build a bridge for the same set of messages as we package when building from source and installing dependencies via rosdep? (question form @nuclearsandwich above)
- Is there another goal that gets accomplished by placing them in the source package.xml?

**Comment by dirk-thomas** [comment]
> Adding these dependencies explicitly in the package.xml means that users will default to build a bigger bridge for messages that are not required and that they may not use.

Can you clarify the workflow you imagine? The set of built packages is defined by what is in the workspace. the way the dependencies are described (group or explicit) won't change anything.

That is why I mention "with a comment" above. These extra dependencies should have a comment in the manifest that they are only being declared there to support common messages.

> As pointed out above now that we have the group dependencies, listing these dependencies in the source package.xml will duplicate information without changing behavior for the source case.

For packaging we need the explicit one. That is the reason why you added them in the release repo in the first place.

> Is it to avoid having any patch for this package in the debian case?

Yes, that was the main motivation.

> Do we want users to build a bridge for the same set of messages as we package when building from source and installing dependencies via rosdep?

Without defining what workflow we are talking about here it doesn't make a different how the dependencies are declared since the only thing that matters is what packages are in the workspace.

---

I am fine keeping the additional dependencies in the release repo if you prefer that. I only raised the question to clarify why it is preferred to have them in the release repo rather then in the upstream repo. In general I think the upstream repo is preferred *until* there is a good reason advocating for the release repo.

**Comment by mikaelarguedas** [comment]
> Can you clarify the workflow you imagine? The set of built packages is defined by what is in the workspace. the way the dependencies are described (group or explicit) won't change anything.

Workflow:
- in a ROS 2 workspace with custom msgs
- clone ros1_bridge repository
- rosdep install ...
  - installs the dependencies listed in the cloned package.xml
- build the bridge

Result:
- the resulting bridge can bridge the messages listed in the package.xml + all custom messages in the sourced workspaces

With this current patch it will not build factories or be able to bridge other message packages if they are not on the system / sourced workspaces (including the ones listed in this PR description).
With the suggested alternative, it will build factories and be able to bridge all packages listed in this PR description.

On the + side: people get a more capable bridge by default
On the - side: this will take more resources and time to build and result in a bigger footprint on the system.

---

> For packaging we need the explicit one. That is the reason why you added them in the release repo in the first place.

> I am fine keeping the additional dependencies in the release repo if you prefer that. I only raised the question to clarify why it is preferred to have them in the release repo rather then in the upstream repo. In general I think the upstream repo is preferred until there is a good reason advocating for the release repo.

I think we agree that listing these deps is needed only for packaging and that it shouldn't change anything for source users as long as they build ROS2 and the bridge following the wiki instructions.

So I guess the decision-making question is:
Should changes related only to packaging and not impacting source builds be hosted on the upstream or the release repo?

My gut feeling was that if it was not related to building the package from source it should be in the release repo. But I also see the value of having a single place to store and maintain all the relevant information.

Whichever we decide I'm fine adopting it for this repository as well as the other ros2 release repositories.

**Comment by dirk-thomas** [comment]
> Workflow:

Thanks for clarifying.

> Result:

Agreed.

> On the + side: people get a more capable bridge by default
> On the - side: this will take more resources and time to build and result in a bigger footprint on the system.

As long as the user only cares about support for the custom msgs I agree that applying the patch in the release repo sounds like the better choice. I just don't think this is a very likely scenario (that the user starts without an underlay and uses rosdep to install the necessary packages).

The case which I was wondering about is when a user has used a "prebuilt" bridge. And then wants to additionally support custom messages. It looks like that the group dependency will be sufficient in that case to ensure that a custom built bridge will at least support the same same as the "prebuilt" bridge because the prebuilt bridge will already ensure that the additional message packages declared in the release repo are present. So the user won't face the problem of "loosing" functionality in the custom built bridge compared to the prebuilt one.

So I am fine keeping the patch in the release repo then.

**Comment by mikaelarguedas** [comment]
> The case which I was wondering about is when a user has used a "prebuilt" bridge. And then wants to additionally support custom messages. It looks like that the group dependency will be sufficient in that case to ensure that a custom built bridge will at least support the same same as the "prebuilt" bridge because the prebuilt bridge will already ensure that the additional message packages declared in the release repo are present. So the user won't face the problem of "loosing" functionality in the custom built bridge compared to the prebuilt one.

Oh yeah that's a good point I didnt think of the potential loss of functionality vs the prebuilt bridge :+1:. Thanks for explaining.

Then I'll move forward and merge this PR as is once it's approved.

**Review by dirk-thomas — APPROVED** [review]

**Comment by dirk-thomas** [comment]
Wouldn't it be better to apply a patch to add the message packages rather then overwrite the `package.xml` file?

**Comment by mikaelarguedas** [comment]
> Wouldn't it be better to apply a patch to add the message packages rather then overwrite the package.xml file?

Oh yeah it would definitely be better.
Do you mean by committing the patches directly on the release/foo/bar branch after running bloom? or is there a way to provide just a patch file on the `master` branch that will be applied ? (I like the things stored in the default branch for visibility)

**Comment by dirk-thomas** [comment]
Maybe @wjwwood can comment if there is a "better" way which is "visible" as well as avoid blindly overwriting the upstream file.

**Comment by wjwwood** [comment]
A patch in the release branch is the earliest patch that gets propagated forward and is not a whole file patch. The "patches" dir that gets applied to the "upstream" branch is only a overwriting file kind of "patch".

**Comment by mikaelarguedas** [comment]
Sounds good.
I'll merge this as is and once I get all the other patches propagated I'll come to revisit the package.xml files overwrite and make them actual release patches.

Thank you both for the input :+1:

**Pull request merged**
